sonar-core/src/main/java/org/sonar/core/platform/PluginLoader.java

```diff
@@ -0,0 +1,192 @@
+      }
+      if (Strings.isNullOrEmpty(info.getBasePlugin())) {
+        // The plugins that extend other plugins can only add some files to classloader.
+        // They can't change ordering strategy.
+        def.setSelfFirstStrategy(info.isUseChildFirstClassLoader());
+      }
+    }
+    return classloadersByBasePlugin.values();
+  }
+
+  /**
+   * Step 2 - create classloaders with appropriate constituents and metadata
+   */
+  private void buildClassloaders(Collection<ClassloaderDef> defs) {
+    ClassloaderBuilder builder = new ClassloaderBuilder();
+    for (ClassloaderDef def : defs) {
+      builder
+        .newClassloader(def.getBasePluginKey(), getClass().getClassLoader())
+        .setExportMask(def.getBasePluginKey(), def.getMask())
+        .setLoadingOrder(def.getBasePluginKey(), def.isSelfFirstStrategy() ? SELF_FIRST : LoadingOrder.PARENT_FIRST);
+      for (File file : def.getFiles()) {
+        builder.addURL(def.getBasePluginKey(), fileToUrl(file));
+      }
+    }
+    Map<String, ClassLoader> classloadersByBasePluginKey = builder.build();
+    for (ClassloaderDef def : defs) {
+      ClassLoader builtClassloader = classloadersByBasePluginKey.get(def.getBasePluginKey());
+      if (builtClassloader == null) {
+        throw new IllegalStateException(String.format("Fail to create classloader for plugin [%s]", def.getBasePluginKey()));
+      }
+      def.setBuiltClassloader(builtClassloader);
+    }
+  }
+
+  /**
+   * Step 3 - instantiate plugin instances ({@link Plugin}
+   *
+   * @return the instances grouped by plugin key
+   * @throws IllegalStateException if at least one plugin can't be correctly loaded
+   */
+  private Map<String, Plugin> instantiatePluginInstances(Collection<ClassloaderDef> defs) {
+    // instantiate plugins
+    Map<String, Plugin> instancesByPluginKey = new HashMap<>();
+    for (ClassloaderDef def : defs) {
+      // the same classloader can be used by multiple plugins
+      for (Map.Entry<String, String> entry : def.getMainClassesByPluginKey().entrySet()) {
+        String pluginKey = entry.getKey();
+        String mainClass = entry.getValue();
+        try {
+          instancesByPluginKey.put(pluginKey, (Plugin) def.getBuiltClassloader().loadClass(mainClass).newInstance());
+        } catch (UnsupportedClassVersionError e) {
+          throw new IllegalStateException(String.format("The plugin [%s] does not support Java %s",
+            pluginKey, SystemUtils.JAVA_VERSION_TRIMMED), e);
+        } catch (Exception e) {
+          throw new IllegalStateException(String.format(
+            "Fail to instantiate class [%s] of plugin [%s]", mainClass, pluginKey), e);
+        }
+      }
+    }
+    return instancesByPluginKey;
+  }
+
+  public void unload(Collection<Plugin> plugins) {
+    for (Plugin plugin : plugins) {
+      ClassLoader classLoader = plugin.getClass().getClassLoader();
+      if (classLoader instanceof Closeable && classLoader != getClass().getClassLoader()) {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Change this comparison to use the equals method.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1698)